Add a new --ignore-file flag to docker build

vdemeester opened the pull request:
This makes it possible to use something else than .dockerignore as a ignore file. Should fix #12886.

```bash
$ docker build -t foo/bar --ignore-file mydockerignore .
# […]
```

Questions/todos : 
* [x] <del>support multiple files ? (`--ignorefile f1 --ignorefile f2` and merge the content naively)</del>
* [ ] update documentation

/cc @thaJeztah @duglin @tiborvass 

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

duglin commented:
Given how many people have asked for this, if we can do it nicely, then I'm in favor of it.

I don't think we need to support multiple files. Once we have a single file done we can explore adding it if people scream for it, but I'm not sure I see a lot of value yet.  Should be easy to add though.

thaJeztah commented:
I agree with @duglin; baby steps, and just having this is already a good start :-)

thaJeztah commented on api/swagger.yaml:
```diff
@@ -4321,6 +4321,11 @@ paths:
           default: "Dockerfile"
+        - name: "ignorefile"
+          in: "query"
+          description: "Path within the build context to the `dockerignore`."
```
i.e., missing _file_ at the end

thaJeztah commented:
Design looks good to me

vdemeester commented:
oh I guess it's breaking the cache & co… hum… I missed something somewhere :joy:

tianon commented:
Any particular reason we can't have both `--ignore` _and_ `--ignore-file`? :innocent:

vdemeester commented:
@tianon what do you mean ?

tianon commented:
So I could just `docker build --ignore .git` rather than writing a whole new file :innocent:

tianon commented:
Also, can this be specified more than once?

Right now, if I wanted to just "build this thing, but also ignore `.git`", I need to copy the entire `.dockerignore` verbatim to add a single line (no concept of includes there either), and then keep it in sync over time (for example, I've cloned someone else's repo and want to ignore extra stuff I've got locally that upstream doesn't care about -- the types of things we tell people to put in `.git/info/excludes` rather than `.gitignore` as concrete examples).

vdemeester commented:
@tianon I hear you :angel: I think we should go step by step, and I would follow up this one with proposal on a `--ignore` and why not support of multiple `--ignore-file` too.

thaJeztah commented:
@vdemeester agreed; was discussing this with @tianon already; this is a good start, and I'm open to future improvements

vdemeester commented:
yup, and I really really like the idea of `--ignore .git` for example, that could help a lot of workflow quite a bit 👼

justincormack commented:
@tianon you don't really want multiple ignores we want to be able to specify multiple bits of context, additive not subtractive. I will do a PR at some point.

justincormack commented on api/swagger.yaml:
```diff
@@ -4321,6 +4321,11 @@ paths:
           default: "Dockerfile"
+        - name: "ignorefile"
+          in: "query"
+          description: "Path within the build context to the `dockerignore` file."
```
should be `.dockerignore` I think

justincormack commented on builder/dockerignore.go:
```diff
@@ -26,8 +26,8 @@ type DockerIgnoreContext struct {
 // files, instead handle a list of files to be excluded from the context.
-func (c DockerIgnoreContext) Process(filesToRemove []string) error {
-	f, err := c.Open(".dockerignore")
+func (c DockerIgnoreContext) Process(fileName string, filesToRemove []string) error {
```
Can we call this `ignoreFile`, its very confusing to have to work out what `fileName` refers to.

vdemeester committed:
Add a new --ignore-file flag to docker build

This makes it possible to use something else than .dockerignore as a
ignore file.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
Rebase and @justincormack comments taken into account 👼

duglin commented on builder/dockerignore.go:
```diff
@@ -26,8 +26,8 @@ type DockerIgnoreContext struct {
-	f, err := c.Open(".dockerignore")
+func (c DockerIgnoreContext) Process(ignoreFile string, filesToRemove []string) error {
+	f, err := c.Open(ignoreFile)
 	// Note that a missing .dockerignore file isn't treated as an error
```
I'm not sure this comment/logic is true in the case of someone being explicit about it.  I would prefer if we threw an error if we were told to use a file and that file didn't exist.  I think it should only silently ignore the "not found" error in the default case.

vdemeester replied:
This would mean making the server side know what the default is (i.e. `.dockerignore`), I think it make sense, even just for retro-compatibility API-wise 👼

thaJeztah replied:
Good one; yes (without looking at how to accomplish it), I think it someone _specified_ a `.dockerignore` file, and it doesn't exist, that should be an error. And if no file was specified, it should not be an error.

But, now sure if that can be achieved :blush:

duglin replied:
One way to know what's going on is to use an empty string to mean "not specified" - then in Process() if its "" you know you can default to .dockerignore and ignore the error if the file isn't there.

duglin commented on client/image_build.go:
```diff
@@ -94,6 +94,7 @@ func (cli *Client) imageBuildOptionsToQuery(options types.ImageBuildOptions) (ur
 	query.Set("cgroupparent", options.CgroupParent)
 	query.Set("shmsize", strconv.FormatInt(options.ShmSize, 10))
 	query.Set("dockerfile", options.Dockerfile)
+	query.Set("ignorefile", options.Ignorefile)
```
Related to my previous comment, if we don't send this query parameter at all when the --ignore-file wasn't specified then the server should be able to ignore the "not found" error because it's defaulting.

vdemeester replied:
@duglin so `docker build --ignore-file .dockerignore` would behave differently than `docker build` ?

duglin replied:
yes. I was thinking that `docker build --ignore-file .dockerignore` would do a couple of things:
1 - check for .dockerignore so that we can fail fast if its not there.  Note, it would only do this check if the user specified the --ignore-file flag.
2 - send the appropriate ignore-file query param to the daemon - but only when the user specified a value. Meaning, if they didn't specify the --ignore-file flag then there would be no query param.
3 - the daemon would detect whether or not the query param was specified. If specified, regardless of its value, if the file isn't there then it would generate an error.  If not there then we use the default value but do not generate an error if the file is missing.  This is similar to how we act on the client - no error for a missing file when we're defaulting, but generate an error when the user was explicit.

vdemeester replied:
ok, so this should be what is implemented now 👼

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3177,6 +3177,51 @@ dir`,
+	dockerfile := `
+        FROM busybox
+        ADD . /bla
+		RUN sh -c "[[ -f /bla/src/x.go ]]"
```
funky formatting

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -3177,6 +3177,51 @@ dir`,
+		c.Fatal(err)
+	}
+}
+
```
we should add a test for when the file isn't found - regardless of whether you change things based on my previous comments.

duglin replied:
Thanks for adding the cli test for this. I think we may need one at the "api" level too - meaning one where we call the REST API with a missing dockerignore file since this just tests the client side, not the server's side.

duglin commented:
we'll need to update the docs for this.

LK4D4 commented:
@vdemeester There is still some comments I believe

vdemeester commented:
@duglin updated to take your comment into account 👼, added a tests case on that too

duglin commented on builder/dockerfile/internals.go:
```diff
@@ -624,7 +624,16 @@ func (b *Builder) readDockerfile() error {
+		err = dockerIgnore.Process(ignoreFile, []string{b.options.Dockerfile})
+		if !ignoreErr && err != nil {
+			return err
+		}
```
Is there any reason this "ignoreErr" logic isn't in Process()?  I worry about putting it here because it assumes that the only possible error from Process() is that .dockerignore doesn't exist - which right now is true. But, if at some point in the future another error is returned then the next person to touch this code **might** do something really hacky like start checking for certain errors - or worse, forget to even check at all and introduce a bug.

Its not a big thing, but I would prefer if all of this new logic were moved into Process() so that we can add the ignoreErr logic right at the `os.IsNotExist(err)` line of code and avoid this potential headache later on.  I think the code will be smaller/easier too.

duglin commented on builder/dockerignore/dockerignore.go:
```diff
@@ -43,7 +43,7 @@ func ReadAll(reader io.Reader) ([]string, error) {
 	}
 	if err := scanner.Err(); err != nil {
-		return nil, fmt.Errorf("Error reading .dockerignore: %v", err)
+		return nil, fmt.Errorf("Error reading dockerignore file: %v", err)
```
Can we add the name of the dockerignore file to this error message to help to user when they "fat finger" it?

duglin commented on cli/command/image/build.go:
```diff
@@ -192,7 +201,7 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		if err == nil {
 			excludes, err = dockerignore.ReadAll(f)
-			if err != nil {
+			if !ignoreErr && err != nil {
```
Is this the right spot for this? if we get an os.IsNotExist() above then won't `f` be nil and then the `ReadAll()` will return immediately with `nil, nil`?  It seems as though `ignoreErr` might be needed at all, or you'll want to add it to the `if err == nil {` line (2 lines above) to skip the ReadAll() in that case.  But, it actually seems like that "if" might be better written as `if f != nil {`

duglin replied:
edited comment, was missing a "not"

justincormack commented:
I do think that specifying the files you don't want is totally wrong by the way. Subtractive specification is hard. I write all my docker builds as `tar cf - Dockerfile files I want | docker build -` so I can additively specify files. If `docker build` accepted more than one file for the context life would be much easier and no one would mess around with `.dockerignore` at all, they would just list the files they wanted...

thaJeztah commented:
Well, it's modeled after `.gitignore`. Also, you'd still be able to do this,  by ignoring everything, then add files you _do_ want as not ignored

justincormack commented:
@thaJeztah the use case quoted in the issue is multiple Dockerfiles in one directory. I use these a lot and ignore is absolutely the wrong way to deal with them.

for example, taking one of my things:
`Dockerfile.1` uses `file.1`, `directory.1`, `file.shared`, `directory.shared`
`Dockerfile.2` uses `file.2`, `directory.2`, `file.shared`, `directory.shared`

So I do
```
tar cf - Dockerfile.1 file.1 directory.1 file.shared directory.shared | docker build -
tar cf - Dockerfile.2 file.2 directory.2 file.shared directory.shared | docker build -
```
and thats great. I want to do, in the future
```
docker build -f Dockerfile.1 file.1 directory.1 file.shared directory.shared
docker build -f Dockerfile.2 file.2 directory.2 file.shared directory.shared
```
After this PR I would have to create two more files:
```
ignore.1 < Dockerfile.2 file.2 directory.2
ignore.2 <  Dockerfile.1 file.1 directory.1
```
and then do
```
docker build -f Dockerfile.1 --ignore ignore.1 .
docker build -f Dockerfile.2 --ignore ignore.2 .
```
This is a terrible way to do things, the ignore file for the first build is listing the files for the second build because it is subtractive not additive. If the second build needs more files, I need to add them to the ignore file for the first one. The more builds I have the more I have to list in each file.

In addition if you use `make`, with the additive way the files for the context are the `make` dependencies, so I tend to just write:
```
build.1: Makefile.1 file.1 file.shared directory.shared
       tar cf - $^ | docker build -
```
which could be the simpler `docker build $^` with multi file contexts.

duglin commented:
@justincormack what's interesting about your use case is that its basically asking for a new Dockerfile command to list the files to be included in the build context, e.g.:
```
```

duglin commented:
@justincormack what's interesting about your use case is that its basically asking for a new Dockerfile command to list the files to be included in the build context. Then you wouldn't need to jump thru the tar steps you do - it would make the Dockerfile more self-describing.

justincormack commented:
I guess you could put it in the Dockerfile, hadn't thought of that.

On 8 Jan 2017 14:33, "Doug Davis" <notifications@github.com> wrote:

> @justincormack <https://github.com/justincormack> what's interesting
> about your use case is that its basically asking for a new Dockerfile
> command to list the files to be included in the build context. Then you
> wouldn't need to jump thru the tar steps you do - it would make the
> Dockerfile more self-describing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/29405#issuecomment-271154743>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdcPFJZVPZR4uDW3sWmTpb3MSzXTc5Zks5rQPO9gaJpZM4LM6R8>
> .
>

thaJeztah commented:
> This is a terrible way to do things, the ignore file for the first build is listing the files for the second build because it is subtractive not additive. If the second build needs more files, I need to add them to the ignore file for the first one. The more builds I have the more I have to list in each file.

It really depends on the use case; if the only difference between Dockerfile1 and Dockerfile2 is to have a single directory _excluded_, having to specify everything else is terrible.

I think allowing multiple paths to be set for the build-context is slightly orthogonal to this PR; this PR just brings having a second Dockerfile "on parity" with the "main" Dockerfile.

I don't think both options are mutually exclusive though

> what's interesting about your use case is that its basically asking for a new Dockerfile command to list the files to be included in the build context. Then you wouldn't need to jump thru the tar steps you do - it would make the Dockerfile more self-describing.

Yes, I think we've discussed such options before, i.e., have the client only send whats actually _needed_ to the daemon. Previous discussions were more about doing so automatically, but having a way to explicitly set what's needed is probably an easier solution.

I do think something like that would make sense, and makes it easier to use, as having to manually specify all paths to include during build makes it quite hard to use ("to build this Dockerfile, specify the following options during `docker build` ...."). Ideally build should be just `docker build -t foo .` in most cases.

justincormack commented:
@thaJeztah no, because when you make changes, eg you refactor your code, you have to refactor a negative, which basically means unrelated code. As you see from my example, the ignore file for build 1 only refers to files that are used in build 2, so if you change build 2, you have to change ignore file 1. This means when you try to split your code, eg split these into different directories, the commit history is horribly entangled.

The model for `docker build` is not `git`, it is `go build`. `go build` takes a directory or a list of files. It does not have an `--exclude` flag, nor does it have an exclude file. Sure this might make your command line shorter in extremely rare situations, but you will not get that added to `go` as the design is wrong: build processes specify what they are built from, not what they are not built from. This is the whole basis of dependency management and analysis.

This change is not actively harmful, it is just adding bloat by taking one existing design decision to a really complex extreme, instead of changing a much simpler part of the design which would make this totally unnecessary.

justincormack commented:
I opened #30101 to discuss alternatives

LK4D4 commented:
I see that this is in code-review, but discuss looks like design-review.
@vdemeester Should we proceed with review?

vdemeester commented:
@LK4D4 maybe it should go back to design-review if there is no consensus 😅

LK4D4 commented:
@vdemeester still @justincormack's proposal was proven to be undoable. Maybe we should discuss on maintainers meeting.

thaJeztah commented:
We discussed this in the maintainers meeting, and think that this is not the right approach; having the builder selectively upload content that is used, will probably be a better solution, and something that is actively looked in to

vdemeester commented:
ok :)

thaJeztah commented:
sorry :cry: :smile:

tzickel commented:
I have encountered this issue just today as well. I've spent time reading all of this discussion and most of the related PR.

My opinion is that there lots of radical ways to improve building in general, but today we have a very nice and simple way to declare some files to be ignored. There is just one glaring issue, while we can name the Dockerfile however we want (and thus allow building different dockers from the same directory), we cannot name the .ignorefile however we want and thus most of us end up sending tons of stuff to the context, that we don't need thus loading the network + build times unnecessary.

I do not understand why a simple PR to just add one option (which should have been done with -f), --ignore-file (even if it's one) conflict with future work on adding other ways to change the context (I assume that due to backwards compability .dockerignore will always be an option and thus people will always want the option to rename it).

dvalentiate commented:
Has anything come along to replace the proposed functionality of this PR?

Not having `--ignore-file` is making life difficult for the teams I'm working with.

thaJeztah commented:
@dvalentiate the issue associated with this PR is https://github.com/moby/moby/issues/12886, feel free to provide input/suggestions there
